### DISCUSSION TITLE
Python の formatter / linter

### DISCUSSION THREAD

**Discussion by habvi** [42-pong/42-pong#6] [backend]
色々ありそうです
他にもあれば
今までは人にお勧めされて black & isort を使ってたけど良かったです

### formatter
- black
- autopep8
- yapf

### linter
- flake8
- pylint

### コード成形ツール
- autoflake
- isort

### 参考
- [Pythonのコードフォーマッターについての個人的ベストプラクティス](https://qiita.com/sin9270/items/85e2dab4c0144c79987d)
- [VSCodeでPython書いてる人はとりあえずこれやっとけ〜](https://qiita.com/nanato12/items/ddf26487eb30714251c3)

**Comment by ayase-mstk** [comment]
早速起票してくださりありがとうございます!
僕もいままでblackでフォーマッターにかけていたくらいなので助かります！

linterはあまり使ったことないのですが、どう使うのがいいと思いますか？
思いつくのは、PRのマージチェックの時に走らせてエラーが出たらマージできないようにするくらいなのですが、これだと結局formatterにかければいいだけだとも思いますし、他によい使い方などあったりしますかね？

あと以前timesでおっしゃっていた型チェックのできるmypyなども使うか決めないとですね。pylanceもいいみたいですが、vscodeでしか使えないのが難点です。。

#### 参考
- [Pythonで中~大規模開発をする際のススメ5選](https://qiita.com/ForestMountain1234/items/70a499b1c00175b77407)

**Comment by habvi** [comment]
コメントありがとうございます！

私も丁度 linter だけ使ったことがなかったです…
linter にしかできないことがあるのかと思っていましたが、black は PEP8 にならって format してくれるようですね
そしたら linter でしかできないことがちょっと不明なのでもう少し調べてみたいところです

pylance 初めて知りました！確かに vscode 依存なのは微妙なところですね…
mypy は python 書いてる人にもお勧めされたことがあるので、良いんじゃないかと思います！
型ヒント、ぜひ採用してみたいです…！

**Comment by ayase-mstk** [comment]
改めて調べてみたところ、大まかにフォーマッターとリンターはこんな感じにできることの違いがあるようです。

|項目 | Formatter（フォーマッター） | Linter（リントツール）
|-- | -- | --
コードの見た目を整える | できる | できない
|インデント・改行・空白の調整 | できる | できない
|PEP 8のスタイルに従う | できる | できる
|未使用の変数・インポートの検出 | できない | できる
|命名規則のチェック | できない | できる
|冗長なコードの警告 | できない | できる
|複雑な構造の指摘 | できない | できる
|パフォーマンスの警告 | できない | できる
|セキュリティリスクの指摘 | できない | できる
|未使用コードの削除 | できない | できる
|構文エラーの検出 | できない | できる（ほとんどのツール）
|コード構造の品質チェック | できない | できる

- **フォーマッター**は、コードのスタイルや見た目を整えるためのツールであり、コード自体の意味や品質には干渉しません。
- **リントツール**は、コーディングのベストプラクティスやコード品質のチェック、セキュリティ、パフォーマンスの問題点などを指摘するのに特化しています。

-> コードのパフォーマンスだったり、未使用変数などの修正はフォーマッターにはできないのでリンターを使う意義はあるかもしれません。


## Ruff
[Pythonの Ruff (linter) でコード整形もできるようになりました](https://qiita.com/ciscorn/items/bf78b7ad8e0e332f891b)
[設定](https://zenn.dev/tk_resilie/articles/python_ruff_setup)
Rust製の[こちら](https://docs.astral.sh/ruff/)のツールを使えば、Flake8 + Black + isortでやっていたことを一つで代替できそうなのでいいかもしれません。

**Reply by habvi** [reply]
ありがとうございます！
Ruff、聞いたことはあったのですが使ったことはなく、記事読んだら色々変わってそうでした。これだけで良さそうな気がしますね！

**Comment by habvi** [comment]
Ruff について
PEP8 は 1 行の幅を 79 文字と定めているらしく、設定した方が良いかも？
https://docs.astral.sh/ruff/tutorial/#configuration

ruff.toml
```
# Set the maximum line length to 79.
line-length = 79
```

**Comment by ayase-mstk** [comment]
結論
linter, formatter -> ruff
type checker -> mypy
